src/main/java/org/embulk/gradle/embulk_plugins/ProspectiveDependencies.java

```diff
@@ -0,0 +1,368 @@
+                throw new GradleException("A tuple of [" + overlapping.getKey().toString() + "] is duplicated unexpectedly.");
+            }
+        }
+        for (final Map.Entry<VersionlessDependency, String> runtimeOnly : runtimeOnlyVersionMap.entrySet()) {
+            if (null != prospectiveDependencies.put(runtimeOnly.getKey(), VersionScope.runtime(runtimeOnly.getValue()))) {
+                throw new GradleException("A tuple of [" + runtimeOnly.getKey().toString() + "] is duplicated unexpectedly.");
+            }
+        }
+        return new ProspectiveDependencies(prospectiveDependencies);
+    }
+
+    @Override
+    public Iterator<ScopedDependency> iterator() {
+        return this.scopedDependencies.iterator();
+    }
+
+    @Override
+    public String toString() {
+        return this.dependencies.toString();
+    }
+
+    private static class Builder {
+        Builder() {
+            this.compileVersionMap = null;
+            this.runtimeVersionMap = null;
+            this.compileException = null;
+            this.runtimeException = null;
+        }
+
+        Builder addCompileClasspath(final Set<ResolvedArtifact> compileClasspath) {
+            if (this.compileVersionMap != null || this.compileException != null) {
+                throw new IllegalStateException(
+                        "ProspectiveDependencies.Builder.addCompileClasspath is called twice unexpectedly.");
+            }
+            try {
+                this.compileVersionMap = buildVersionMapFromResolvedArtifacts(compileClasspath);
+            } catch (final UnexpectedDependencyException ex) {
+                this.compileException = ex;
+            }
+            return this;
+        }
+
+        Builder addRuntimeClasspath(final Set<ResolvedArtifact> runtimeClasspath) {
+            if (this.runtimeVersionMap != null || this.runtimeException != null) {
+                throw new IllegalStateException(
+                        "ProspectiveDependencies.Builder.addRuntimeClasspath is called twice unexpectedly.");
+            }
+            try {
+                this.runtimeVersionMap = buildVersionMapFromResolvedArtifacts(runtimeClasspath);
+            } catch (final UnexpectedDependencyException ex) {
+                this.runtimeException = ex;
+            }
+            return this;
+        }
+
+        ProspectiveDependencies build() throws GradleException {
+            if (this.compileException != null || this.runtimeException != null) {
+                throw new GradleException();  // TODO:
+            }
+            if (this.compileVersionMap == null || this.runtimeVersionMap == null) {
+                throw new NullPointerException(
+                        "ProspectiveDependencies is built without compileClasspath or runtimeClasspath unexpectedly.");
+            }
+            return ProspectiveDependencies.buildFromVersionMaps(this.compileVersionMap, this.runtimeVersionMap);
```

Comment:
I'm wondering why we need to call upstream to parent here while we can do it in Builder, is it for easier to test?

Reply:
Good point. Yeah, one big reason is for easier testing, and the other reason is because I didn't want to make the `Builder` class so big as it's an inner class.